### PR TITLE
General maintenance for Lathes

### LINKED ISSUE #81192: Autolathe can't print items with custom materials
## Issue Summary

I tried to print toolboxes with different materials, popup msg said "Not enough materials" (or smth like this) and autolathe stuck in printing animation.

### LINKED ISSUE #81236: Lathes not printing multiple items correctly
Server-Side Code Modifications: https://tgstation13.org/parsed-logs/sybil/server_side_modifications.dm

Round ID: [223218](https://scrubby.melonmesa.com/round/223218)
BYOND Version: 515.1626
Server revision compiled on: 
Master commit: [943f38bf7c5f9c048cc785deb0c537d57ee6ba77](https://www.github.com/tgstation/tgstation/commit/943f38bf7c5f9c048cc785deb0c537d57ee6ba77)
The following pull requests are currently test merged:
[#81089: 'Temp keybinds patch for 515.1628' by MrStonedOne at commit 45649ea7a5](https://www.github.com/tgstation/tgstation/pull/81089)
[#81120: 'Comprehensive cleanup of storage datum, replaces the weakrefs with just refs (because they were managed already)' by MrMelbert at commit bb98f3e02a](https://www.github.com/tgstation/tgstation/pull/81120)

Local commit: 97bde235abf11a67c68b2db76a092a0d613bcbd1
TGS version: /tg/station 13 Server v3.2.6.2
DMAPI version: 3.2.1.3

I tried to print 5 large beakers at the service lathe, and it only printed 1.  There were sufficient materials.  It was the same whether I set the custom amount to 5, or I hit the button that said x5.

### LINKED ISSUE #81243: Lathe's just reload their whole interface every second time I print an item
## Issue Summary

So I print an item and another item. Lathe window gets the loading suitcase thing and goes smaller for a moment, then big again, making the list of items scrolled up again so I have to scroll it back down again.

It happens when new designs arrive too, but there it's understandable at least.

Used to think it was random, but with the non-multiple printing part it was really apparent it's on every second print.

## Reproduction:

Print an item on the lathe
Print another
See the window reload

### PR DESCRIPTION
**Pull request by SyncIt21**
## About The Pull Request
1. **Qol Stuff**
   - Screentips & examines for screwdriver, crowbar acts, multiool & wirecutter Also for Alt click
   - Techfabs can now also use the Mouse drag functionality to set drop target for items
   - Lathe printing animation now plays on loop instead of just flicking once till printing is finished for more visual feedback

2. **Code Improvements**
    - Merged `start_making()` with `do_make_item()`. That proc was like only 3 lines long and used only in 1 place so let's just move that code to `ui_act()`
    - Merged `user_print_item_id()` with `ui_act()`. Again was used only in 1 place so let's just move that code in to save some proc overhead
    - Sets `processing_flags` for autolathe to `NONE` cause we don't use `process()`
    - Autodocs vars such as `hacked` , `shocked` etc & procs
    - `maxmult` is now computed client side saving backend bandwidth, `construction_time` is removed from lathes which did not use it
    - Removed all usages of lathe taxes and their related vars, removed engineering lathe no tax from ice moon, replaced with normal engineering lathe

3. **Fixes**
    -  Lathe sheet insertion animations are now linked & work again for all material types inserted via remote silo/local storage, silver/titanium/plastic all play the same animation(that is `protolathe_shiny` overlay). Other materials have their own respective overlays
    - Fixes #81243. Calling `update_static_data_for_all_viewers()` is too expensive for the UI. We should instead use `SStgui.update_uis(src)` which will report the `busy` status to the UI more immediatly
    - Fixes #81236. Some problems with the params passed to the timer callback. It should now print the correct number of requested items
    - Fixes #81192. `design.materials` would runtime for custom material items as they were list of texts not materials. We have to pass our manually parsed list of materials for an specific item to ensure they are set & used correctly. Same fixes apply for techfabs as well

## Changelog
:cl:
qol: adds screentips & examines for screwdriver & crowbar acts & alt click.
qol: techfabs can now use the mouse drop functionality to set drop target.
qol: lathe printing animation plays on loop while printing rather than flicking once for more visual feedback
fix: lathe sheet insertion animations are now linked & work again for all material types inserted via remote silo/local storage
fix: printing custom materials items from autolathe works again.
fix: printing multiple items from lathes will actually print that correct quantity of items requested.
fix: printing items the 2nd time around from lathes won't cause the UI to reload each time.
code: autodoc for some vars & procs, merges procs.
refactor: Optimized code for autolathe & techfabs in general. Report bugs on github
/:cl:

### CODE REVIEW

File: code/game/machinery/autolathe.dm [hunk 1]
```diff
@@ -43,17 +43,57 @@
...
+			. += span_notice("[EXAMINE_HINT("Drag")] towards a direction (while next to it) to change drop direction.")
+
+		. += span_notice("Its maintainence panel can be [EXAMINE_HINT("screwed")] [panel_open ? "closed" : "open"].")
+		if(panel_open)
+			. += span_notice("The machine can be [EXAMINE_HINT("pried")] apart.")
+
+/obj/machinery/autolathe/add_context(atom/source, list/context, obj/item/held_item, mob/user)
+	if(isnull(held_item) && drop_direction)
+		context[SCREENTIP_CONTEXT_ALT_LMB] = "Reset"
+		return NONE
```
Comment (MrMelbert) [inline code/game/machinery/autolathe.dm:70]:
```suggestion
		return CONTEXTUAL_SCREENTIP_SET
```

File: code/game/machinery/autolathe.dm [hunk 2]
```diff
@@ -43,17 +43,57 @@
...
+
```
Comment (MrMelbert) [inline code/game/machinery/autolathe.dm:71]:
```suggestion
	if(isnull(held_item))
		return NONE
```
You sanity check for `isnull(held_item)` but only when `drop_direction` is set.

File: code/game/machinery/autolathe.dm [hunk 3]
```diff
@@ -189,7 +238,7 @@
 	build_count = clamp(build_count, 1, 50)
 
 	var/list/materials_needed = list()
-	for(var/datum/material/material as anything in design.materials)
+	for(var/material as anything in design.materials)
```
Comment (MrMelbert) [inline code/game/machinery/autolathe.dm:241]:
```suggestion
	for(var/material in design.materials)
```

File: code/game/machinery/autolathe.dm [hunk 4]
```diff
@@ -43,17 +43,57 @@
...
+	register_context()
+
 /obj/machinery/autolathe/Destroy()
 	materials = null
 	QDEL_NULL(wires)
 	return ..()
 
+/obj/machinery/autolathe/examine(mob/user)
+	. = ..()
+	if(in_range(user, src) || isobserver(user))
```
Comment (MrMelbert) [inline code/game/machinery/autolathe.dm:55]: Early return this?

File: code/game/machinery/autolathe.dm [hunk 5]
```diff
@@ -43,17 +43,57 @@
...
+			. += span_notice("[EXAMINE_HINT("Alt-click")] to reset.")
+		else
+			. += span_notice("[EXAMINE_HINT("Drag")] towards a direction (while next to it) to change drop direction.")
+
+		. += span_notice("Its maintainence panel can be [EXAMINE_HINT("screwed")] [panel_open ? "closed" : "open"].")
+		if(panel_open)
+			. += span_notice("The machine can be [EXAMINE_HINT("pried")] apart.")
+
+/obj/machinery/autolathe/add_context(atom/source, list/context, obj/item/held_item, mob/user)
+	if(isnull(held_item) && drop_direction)
```
Comment (MrMelbert) [inline code/game/machinery/autolathe.dm:68]: You sanity check for `isnull(held_item)`, but only when `drop_direction` is set. 

So this proc will runtime if `held_item` is `null`, but `drop_direction` is not set, later down the line.

Comment (MrMelbert) [reply code/game/machinery/autolathe.dm:68]: Also, ostensibly, you can alt-click on things while holding items. So this is isn't necessarily accurate.

**Comment by TJatPBnJ** [comment]
This PR does not completely fix #81236 or #81243, as protolathes, circuit imprinters, and techfabs are unchanged.

File: code/game/machinery/autolathe.dm [hunk 6]
```diff
@@ -215,53 +269,63 @@
...
-	var/charge_per_item = total_charge / build_count
-
-	var/total_time = (design.construction_time * design.lathe_time_factor * build_count) ** 0.8
-	var/time_per_item = total_time / build_count
-	start_making(design, build_count, time_per_item, material_cost_coefficient, charge_per_item)
-	return TRUE
-
-/// Begins the act of making the given design the given number of items
-/// Does not check or use materials/power/etc
-/obj/machinery/autolathe/proc/start_making(datum/design/design, build_count, build_time_per_item, material_cost_coefficient, charge_per_item)
```
Comment (GoldenAlpharex) [inline code/game/machinery/autolathe.dm:236]: I don't think that moving contents from a separate proc into one that's already over a hundred lines long is a code improvement. That just makes that long proc, even longer, and if we somehow wanted to have another way to start making an item from the autolathe without using the UI, we no longer have that option anymore.

This feels like the inverse of a code improvement, imo.

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]

File: code/game/machinery/autolathe.dm (hunk 6 above)
Comment (SyncIt21) [reply code/game/machinery/autolathe.dm:236]: It's a code improvement in the sense we have don't have a chain of proc calls saving us a little overhead 

We should use procs only if that code is being used in atleast more than 1 place which wasn't the case here 

> and if we somehow wanted to have another way to start making an item from the autolathe without using the UI

There is no such use case here so I don't see a need for that, We don't plan code for a future project

**Comment by SyncIt21** [comment]
> This PR does not completely fix #81236 or #81243, as protolathes, circuit imprinters, and techfabs are unchanged.

I see the problem is more extensive than I thought. Will draft and fix them

**Comment by SyncIt21** [comment]
> > This PR does not completely fix #81236 or #81243, as protolathes, circuit imprinters, and techfabs are unchanged.
> 
> I see the problem is more extensive than I thought. Will draft and fix them

All issues addressed for techfabs & autolathe. Would be nice if i got the refactor label as this PR became larger than initially intended

**Comment by SuperNovaa41** [comment]
Closes https://github.com/tgstation/tgstation/pull/81267

**Comment by SyncIt21** [comment]
> Closes #81267

Thank you. can you close that PR for me? the author is acting childish and not listening

**Comment by Iajret** [comment]
Cant test this on TG codebase at the moment, but it seems, department stripes does not appear on protolathes. They all are plain gray.

**Comment by Iajret** [comment]
but at current state it does close https://github.com/tgstation/tgstation/issues/81132 and closes https://github.com/tgstation/tgstation/issues/80853

**Comment by Iajret** [comment]
Also, protolathes seem to eat all available power from APCs and doesn't even update their status afterward so you are left with something like this
image

**Comment by SyncIt21** [comment]
> Cant test this on TG codebase at the moment, but it seems, department stripes does not appear on protolathes. They all are plain gray.

Thanks i'm looking into it

> Also, protolathes seem to eat all available power from APCs and doesn't even update their status afterward so you are left with something like this

Power changes for all lathes were changed in #80839 and to better understand the issue refer #81167. It's not a bug apparently

**Comment by Iajret** [comment]
> It's not a bug apparently

Power consumption isn't. But the fact that APCs does not update their charging status - is. You are right though. Probably #81198 introduced the issue. Disregard it then, i think

**Comment by SyncIt21** [comment]
Stripes appear again

File: code/game/machinery/autolathe.dm [hunk 7]
```diff
@@ -283,26 +337,44 @@
...
 
-/obj/machinery/autolathe/screwdriver_act_secondary(mob/living/user, obj/item/tool)
-	if(default_deconstruction_screwdriver(user, "autolathe_t", "autolathe", tool))
-		return ITEM_INTERACT_SUCCESS
+/obj/machinery/autolathe/AltClick(mob/user)
+	. = ..()
+	if(!drop_direction || !can_interact(user))
+		return
+	if(busy)
+		balloon_alert(user, "printing started!")
```
Comment (Ghommie) [inline code/game/machinery/autolathe.dm:374]:
```suggestion
		balloon_alert(user, "busy printing!")
```

File: code/modules/research/destructive_analyzer.dm [hunk 8]
```diff
@@ -14,16 +14,29 @@
...
+
+/obj/machinery/rnd/destructive_analyzer/examine(mob/user)
+	. = ..()
+	if(!in_range(user, src) && !isobserver(user))
+		return
+
+	if(loaded_item)
+		. += span_notice("Use [EXAMINE_HINT("LMB")] to remove loaded item inside.")
+	else
+		. += span_notice("An item can be loaded inside via [EXAMINE_HINT("LMB")].")
```
Comment (Ghommie) [inline code/modules/research/destructive_analyzer.dm:39]:
```suggestion
		. += span_notice("[EXAMINE_HINT("Left-Click")] to remove loaded item inside.")
	else
		. += span_notice("An item can be loaded inside via [EXAMINE_HINT("Left-Click")].")
```
I believe `Left-Click` to be faster to process than the acronym.

File: code/modules/research/machinery/_production.dm [hunk 9]
```diff
@@ -305,40 +400,36 @@
...
 
-/obj/machinery/rnd/production/update_overlays()
+/obj/machinery/rnd/production/MouseDrop(atom/over, src_location, over_location, src_control, over_control, params)
 	. = ..()
-
-	if(!stripe_color)
+	if((!issilicon(usr) && !isAdminGhostAI(usr)) && !Adjacent(usr))
 		return
+	if(busy)
+		balloon_alert(usr, "printing started!")
```
Comment (Ghommie) [inline code/modules/research/machinery/_production.dm:419]:
```suggestion
		balloon_alert(usr, "busy printing!")
```

File: code/modules/research/machinery/_production.dm [hunk 10]
```diff
@@ -305,40 +400,36 @@
...
-/obj/machinery/rnd/production/examine(mob/user)
+/obj/machinery/rnd/production/AltClick(mob/user)
 	. = ..()
-
-	if(in_range(user, src) || isobserver(user))
-		. += span_notice("The status display reads: Storing up to <b>[materials.local_size]</b> material units.<br>Material consumption at <b>[efficiency_coeff * 100]%</b>.<br>Build time reduced by <b>[100 - efficiency_coeff * 100]%</b>.")
+	if(!drop_direction || !can_interact(user))
+		return
+	if(busy)
+		balloon_alert(user, "printing started!")
```
Comment (Ghommie) [inline code/modules/research/machinery/_production.dm:432]:
```suggestion
		balloon_alert(user, "busy printing!")
```

File: code/modules/research/rdmachines.dm [hunk 11]
```diff
@@ -37,6 +38,46 @@
 	QDEL_NULL(wires)
 	return ..()
 
+/obj/machinery/rnd/examine(mob/user)
+	. = ..()
+	if(!in_range(user, src) && !isobserver(user))
+		return
+
+	. += span_notice("An [EXAMINE_HINT("multitool")] with techweb designs can be uploaded here.")
```
Comment (Ghommie) [inline code/modules/research/rdmachines.dm:46]:
```suggestion
	. += span_notice("A [EXAMINE_HINT("multitool")] with techweb designs can be uploaded here.")
```

**Review by Ghommie — CHANGES_REQUESTED** [review]
So called nits.

**Comment by SyncIt21** [comment]
@Ghommie would be nice if I got the refactor label please?

**Pull request merged**